### ISSUE TITLE
tf.eye is OOM when num_rows is over 10**5

### ISSUE THREAD

**Issue by RyosukeMatsushima** [tensorflow/tensorflow#44748]
<em>Please make sure that this is a feature request. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:feature_template</em>

**System information**
- TensorFlow version (you are using): 2.3.0
- On google colab free GPU
- Are you willing to contribute it (Yes/No): I want to try.

**Describe the feature and the current behavior/state.**
```
element_count = 10**5
eye_tf = tf.eye(element_count, dtype=tf.float32)
```
This code will be OOM. The error is here.

```
ResourceExhaustedError: OOM when allocating tensor with shape[100000,100000] and type float on /job:localhost/replica:0/task:0/device:GPU:0 by allocator GPU_0_bfc [Op:MatrixDiagV3] name: diag
```

**Will this change the current api? How?**
No. 

**Who will benefit with this feature?**
I want to calculate huge array (the num_rows is over 10**8) using tesorflow parallel processing.
This change extend the possibility of tensorflow calculation.

**Any Other info.**

**Comment by ravikyram** [comment]
I have tried in colab with TF 2.3 , nightly version (`2.5.0-dev20201110`) and was able to reproduce the issue. Please, find the gist [here](https://colab.research.google.com/gist/ravikyram/54e5fe7e9285bc103524e5e3958165e7/untitled512.ipynb).Thanks!

**Comment by bhack** [comment]
@RyosukeMatsushima how much memory are you asking for this to Colab GPU? 
https://datascience.stackexchange.com/questions/72870/how-to-find-the-size-of-a-tensor-in-bytes

**Comment by RyosukeMatsushima** [comment]
> @RyosukeMatsushima how much memory are you asking for this to Colab GPU?
> https://datascience.stackexchange.com/questions/72870/how-to-find-the-size-of-a-tensor-in-bytes

@bhack sorry for the late response. Calculate according to your link, memory of eye_tf $ = 10**5**2$elements $\times 4bytes$(float32) $= 40$GB.
Is it possible for tensorflow to calculate a tensor over GPU memory size?. The size of Colab GPU memory is 12GB.
```
!free -h
```
```
  total        used        free      shared  buff/cache   available
Mem:            12G        544M         10G        960K        1.9G         11G
Swap:            0B          0B          0B
```
I'm so glad if tensorflow can calculate huge array over memory.

**Comment by bhack** [comment]
I think you can close this and upvote/follow https://github.com/tensorflow/tensorflow/issues/29840

**Comment by RyosukeMatsushima** [comment]
Thank you!
It seems that this issue will be solved by #29840 and I close this.
